CPU Statistics

deepikabhavnani opened the pull request:
### Description
API to get CPU stats like sleep/deepsleep time, uptime and idle time. These can be used by application to know the CPU Usage runtime.

Preceding PR dependency: #https://github.com/ARMmbed/mbed-os/pull/6821

Status: In progress

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

CC @SenRamakri

c1728p9 committed:
Add API to get idle time

Add the API to get  time sleeping. This is in preparation for CPU
usage support.

Commit:
CPU statistics addition

API to get CPU stats like sleep/deepsleep time, uptime and idle time.
These can be used by application to know the CPU Usage runtime.

deepikabhavnani commented:
Rebased - No PR dependency now

deepikabhavnani commented:
CC @flit

deepikabhavnani commented:
Suggestion from @c1728p9 was to have single low power ticker and idle time can be calculated as sum of sleep and deep sleep timings. Will be compatible even with system having single thread.

c1728p9 commented on rtos/rtos_idle.h:
```diff
@@ -26,6 +26,7 @@
 #define RTOS_IDLE_H
 
 #include <stddef.h>
+#include <stdint.h>
```
Is this header still needed?

deepikabhavnani replied:
Good catch 👍 Will remove it

0xc0170 commented:
Can you review astyle travis check - few style changes requested there. 

We'll have a look why it is not reported here (as +ZZ warnings)

bulislaw commented on hal/mbed_sleep_manager.c:
```diff
@@ -20,13 +20,62 @@
+    return (sleep_time+deep_sleep_time);
+}
+
+uint64_t mbed_uptime(void)
```
`mbed_uptime` is the same as `read_us` can we remove one? or at least make one simply call the other?

0xc0170 commented on hal/mbed_sleep_manager.c:
```diff
@@ -20,13 +20,62 @@
+
+uint64_t mbed_uptime(void)
+{
+#if defined(MBED_CPU_STATS_ENABLED) && defined(DEVICE_LOWPOWERTIMER)
```
isn't this same as read_us ? (should this call that function ?)

deepikabhavnani replied:
Will change

0xc0170 commented on platform/mbed_power_mgmt.h:
```diff
@@ -205,6 +205,32 @@ static inline void system_reset(void)
+ *
+ *  @return  Time spent in sleep
+ */
+uint64_t mbed_time_sleep(void);
```
why was `uint64_t` chosen as a unit to be returned here ? Don't we have time based user defined variables - `us_timestamp_t` - not good fit here? Based on the implementation, I would expect that us timestamp would fit here? As it is, I am missing the units returned (its microseconds after looking at the implementation).

deepikabhavnani replied:
us_timestamp_t is perfect fit, but its declaration is guarded by `DEVICE_LOWPOWERTIMER`, hence will get compilation errors for targets not supporting LP.

deepikabhavnani replied:
@c1728p9 - Can i move `#include "hal/ticker_api.h"` outside `#if DEVICE_LOWPOWERTIMER` define?

c1728p9 replied:
@deepikabhavnani it should be fine to move `#include "hal/ticker_api.h"` outside `#if DEVICE_LOWPOWERTIMER`

bulislaw commented on platform/mbed_power_mgmt.h:
```diff
@@ -205,6 +205,32 @@ static inline void system_reset(void)
     NVIC_SystemReset();
 }
+ 
+/** Provides the time spent in sleep mode, since system is up and running
```
`since system is up and running` that's a bit confusing, can just say from boot? Also we should call out in each of the functions (`@note`) it only works if platform supports LP ticker.

bulislaw commented:
On top of all the statistics changes to the code we need good docs in handbook.

0xc0170 commented on platform/mbed_power_mgmt.h:
```diff
@@ -205,6 +205,32 @@ static inline void system_reset(void)
+ *
+ * @return  System uptime.
+ */
+uint64_t mbed_uptime(void);
```
We have `clock()` in retarget that is similar to this, should that be updated to use this new functionality (clock is implemented with us ticker read that is not ideal).

deepikabhavnani replied:
@0xc0170 - Good point. I believe updating will be good, when we have lp ticker for all targets.
Else we can use lp ticker as default and us ticker as fall back option. If we agree on this, then i will create new PR for this change

0xc0170 replied:
@kjbracey-arm @c1728p9 What do you think? 

For me 👍  - keeping clock as it is if lp ticker is not supported, otherwise use uptime function

c1728p9 replied:
Sounds good to me

kjbracey replied:
`clock()` is supposed to be "CPU run time", not "uptime". What it actually is on mbed OS I think depends on platform config. I believe because it uses us_ticker and doesn't lock it, it's sort of "runtime" if the platform ends up stopping us_ticker during idle. If you can get a real "non-idle" time from the RTOS, that's what clock should really be returning.

The "uptime" is accessible now as `Kernel::get_ms_count()` - that's defined as counting up from 0, and will include any "sleep because of idle" time (because ticks are corrected after idle sleeps).

deepikabhavnani replied:
@kjbracey-arm - Thanks for clarifying, that clock is "run time", and it makes sense to leave it as us_ticker.

deepikabhavnani commented:
> On top of all the statistics changes to the code we need good docs in handbook.

@bulislaw - https://github.com/ARMmbed/Handbook/pull/508. Review/Comments appreciated to make docs good.

Commit:
Addressed review comments

1. LP ticker limiation note
2. Use read_us in mbed_uptime function
3. Doxygen recommendations
4. Use us_timestamp_t instead uint64_t
5. Astyle changes

cmonr commented:
/morph build

Commit:
Updated test to not use event queue.

Test was failing on limited RAM devices, because of shared event queue size.
Updated test to use thread instead of event queue.

0xc0170 commented:
/morph build

Commit:
Added additional read to care of lazyily initialized timer

deepikabhavnani commented:
Test failed on NUCLEO_F746ZG, adding additional read to take care of lazy initialization of lp ticker.

cmonr commented:
@deepikabhavnani Will restart in a bit. Giving PRs targeted for 5.8.5 a chance to complete.

0xc0170 commented:
/morph build

adbridge commented:
Testing has passed but

@0xc0170 @bulislaw @c1728p9 @kjbracey-arm this still needs review approval from 2 of you!

cmonr commented:
FYI, somehow this test build was magic because two other PRs have failed with this update (https://github.com/ARMmbed/mbed-os/pull/6910#issuecomment-391442239) (https://github.com/ARMmbed/mbed-os/pull/6847#issuecomment-391495858).

@deepikabhavnani has a fix ready here (https://github.com/ARMmbed/mbed-os/pull/6993), and it's been locally tested multiple times without failure. Conversely, if master is used, it tends to fail quite often. The issue is possibly due to CI load.

In the future, we should probably consider seriously running multiple test builds back to back when new or updated tests are involved to make sure they get heavy testing on them.
